Added nperiodic_dimensions field (#278)
Materials-Consortia/OPTIMADE: optimade.rst

  - For a 2D surface/slab, periodic on the plane defined by the first and third lattice vectors: :val:`[1, 0, 1]`
  - For a bulk 3D system: :val:`[1, 1, 1]`

nperiodic\_dimensions
~~~~~~~~~~~~~~~~~~~~~

> - **Description**: An integer specifying the number of periodic dimensions in the structure, equivalent to the number of non-zero entries in :property:`dimension\_types`.

Reviewer: Do we really need another 'integer', when the property is defined to be integer?

Also, there is a subtle difference between "sum of entries" and "number of non-zero entries". Do we have a preference, and if so, why?

(I have a preference *against* sum of non-zero entries, because that is just unnecessarily convoluted way to express this, since the zero entries doesn't contribute to the sum...)

> - **Description**: An integer specifying the number of periodic dimensions in the structure, equivalent to the number of non-zero entries in :property:`dimension\_types`.

Reviewer 2: I agree with @rartino

> - **Description**: An integer specifying the number of periodic dimensions in the structure, equivalent to the number of non-zero entries in :property:`dimension\_types`.

Reviewer 3: Consensus has been reached!

- **Type**: integer
- **Requirements/Conventions**:

>   - **Support**: SHOULD be supported by all implementations, i.e., SHOULD NOT be :val:`null`.

Reviewer 2: I guess it's ok since also `dimension_types` is SHOULD, but I would rather make both `MUST`? Or there was a reason for which we said SHOULD? Anyway, this can be treated in a different discussion thread maybe?

Author: I was actually quite surprised at how few of our properties are `MUST`s, (`id`, `type`, `structure_features` only), especially when support from querying on almost all properties is a `MUST`...

Reviewer 3: This comes from a rewrite by @merkys in discussion with @rartino and myself at some earlier point, re-evaluating "required" properties in a response.

Reviewer 2: Still, I'm surprised that `dimension_types` is optional. A Si_2 molecule and a Si crystal might only differ on the `dimension_types`, and they are *very different* things. I would vote to reintroduce this back as a MUST even if I understand this might require further discussion... do you have a link to the issue/PR where this was discussed?

Reviewer: Why are you singling out `dimension_types`? All of the structural properties, `cartesian_site_positions`, `speices_at_sites`, and `species`, (and so on) are SHOULD. 

I suppose I was thinking that we put them as SHOULD now and see how it works out in implementations. If everyone implements them and no one sees any issues with them, we can increase the requirement to MUST.

(It is easier to change properties to mandatory than the other way around, since making mandatory fields non-mandatory means that you are technically breaking backwards compatibility.)

Reviewer 2: I see, fine. But I'll open an issue about this.

>   - **Query**: MUST be a queryable property with support for all mandatory filter features.
>   - The integer value MUST be between 0 and 3 inclusive.

Reviewer: Suggestion for clarity:
```suggestion
  - **Query**: MUST be a queryable property with support for all mandatory filter features.
  - The integer value MUST be between 0 and 3 inclusive and MUST be equal to the sum of the items in the `dimension_types`_ property.
```

>   - The integer value MUST be between 0 and 3 inclusive and MUST be equal to the sum of the items in the `dimension_types`_ property.

Reviewer 3: Just a question: The way of stating that both values (0 and 3) are inclusive here that is the standard English way of writing this? Or are there some "fluff" missing around it?

Author: ```suggestion
  - This property MUST take an integer value from 0 to 3 and MUST be equal to the sum of the items in the `dimension_types`_ property.
```

I think I prefer my original wording, but is this clearer for you?

Reviewer 2: If useful, both would be fine with me

Reviewer 3: The `inclusive` part is integral, it was more that I didn't immediately recognize the sentence structure when using `inclusive`. To be better for non-native English speakers it could simply be a more verbose structure, e.g.:
```suggestion
  - The integer value MUST be between 0 and 3 including both extrema and MUST be equal to the sum of the items in the `dimension_types`_ property.
```
Or something similar, I'm also not loving this structure, so in the end, if it's proper English the first time around, then that's also fine with me. And I then apologize for my lack of mastering the English language :) But happy to learn.

Reviewer: I did not have trouble reading the original formulation and find "including both extrema" a very unusual way of expressing this. But if we really must reformulate it, we'd save characters on: "MUST be 0, 1, 2 or 3". There is also "MUST be a value from 0 to 3", which I think generally is viewed as less ambiguous than "between" and then may not need the 'inclusive' clarification.

Reviewer: We now have two approvals for this PR, but perhaps we should wait for @CasperWA to say if he is happy as it is, or if he wants the "From ... to..." solution, or something else.

Reviewer 2: I like @ml-evs latest suggestion. @CasperWA ?

Reviewer 3: I withdraw my suggested corrections for this, as there is consensus on it :)
